### PR TITLE
Add GraphQL API types and specs

### PR DESCRIPTION
**Pull request by andreslucena**
This PR is the second part of the `decidim-elections` extraction.

In this case, we're extracting the GraphQL API functionality. 

I've changed the type names so we don't have collisions with the Election types. 

Related to #9, as this will allow us to get the Election metadata to use with the Vocdoni SDK

### CODE REVIEW

File: lib/decidim/vocdoni/test/factories.rb [hunk 1]
```diff
@@ -46,6 +46,68 @@
...
+            :attachment,
+            :with_image,
+            attached_to: election
+          )
+        end
+      end
+    end
+  end
+
+  factory :question, class: "Decidim::Vocdoni::Question" do
```
Comment (microstudi) [inline lib/decidim/vocdoni/test/factories.rb:74]:
```suggestion
  factory :vocdoni_question, class: "Decidim::Vocdoni::Question" do
```

File: lib/decidim/vocdoni/test/factories.rb [hunk 2]
```diff
@@ -46,6 +46,68 @@
...
+      answers { 10 }
+    end
+
+    trait :projects do
+      complete
+      answers { 6 }
+    end
+  end
+
+  factory :election_answer, class: "Decidim::Vocdoni::Answer" do
```
Comment (microstudi) [inline lib/decidim/vocdoni/test/factories.rb:105]:
```suggestion
  factory :vocdoni_election_answer, class: "Decidim::Vocdoni::Answer" do
```

File: spec/types/decidim/vocdoni/vocdoni_answer_type_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+require "decidim/api/test/type_context"
+require "decidim/core/test/shared_examples/attachable_interface_examples"
+require "decidim/core/test/shared_examples/traceable_interface_examples"
+
+module Decidim
+  module Vocdoni
+    describe VocdoniAnswerType, type: :graphql do
+      include_context "with a graphql class type"
+
+      let(:model) { create(:election_answer) }
```
Comment (microstudi) [inline spec/types/decidim/vocdoni/vocdoni_answer_type_spec.rb:13]:
```suggestion
      let(:model) { create(:vocdoni_election_answer) }
```

File: spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+module Decidim
+  module Vocdoni
+    # As there's already a GraphQL type called ElectionType (from the module decidim-elections)
+    # we change the name to VodoniElectionType for the external API.
+    # Note that internally there aren't any colition as Ruby has namespaces, meaning that
+    # Decidim::Vocdoni::Election and Decidim::Vocdoni::Election isn't the same thing.
+    describe VocdoniElectionType, type: :graphql do
+      include_context "with a graphql class type"
+
+      let(:model) { create(:election, :published) }
```
Comment (microstudi) [inline spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb:17]:
```suggestion
      let(:model) { create(:vocdoni_election, :published) }
```

File: spec/types/decidim/vocdoni/vocdoni_question_type_spec.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+require "spec_helper"
+require "decidim/api/test/type_context"
+require "decidim/core/test/shared_examples/traceable_interface_examples"
+
+module Decidim
+  module Vocdoni
+    describe VocdoniQuestionType, type: :graphql do
+      include_context "with a graphql class type"
+
+      let(:model) { create(:question) }
```
Comment (microstudi) [inline spec/types/decidim/vocdoni/vocdoni_question_type_spec.rb:12]:
```suggestion
      let(:model) { create(:vocdoni_question) }
```

File: spec/types/integration_schema_spec.rb [hunk 6]
```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+require "decidim/api/test/component_context"
+
+describe "Decidim::Api::QueryType" do
+  include_context "with a graphql decidim component"
+  let(:component_type) { "Vocdoni" }
+  let!(:current_component) { create :vocdoni_component, participatory_space: participatory_process }
+  let!(:election) { create(:election, :published, :finished, component: current_component) }
```
Comment (microstudi) [inline spec/types/integration_schema_spec.rb:10]:
```suggestion
  let!(:election) { create(:vocdoni_election, :published, :finished, component: current_component) }
```

**Review by microstudi — COMMENTED** [review]
This all ok except the same thing as #2, let's namescape factories.

File: spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        it "returns the election's published time" do
+          expect(Time.zone.parse(response["publishedAt"])).to be_within(1.second).of(model.published_at)
+        end
+      end
+
+      describe "blocked" do
+        let(:query) { "{ blocked }" }
+
+        context "when the election's parameters are blocked" do
+          let!(:model) { create(:election, :finished) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb:85]:
```suggestion
          let!(:model) { create(:vocdoni_election, :finished) }
```

File: spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+          it "returns true" do
+            expect(response["blocked"]).to be true
+          end
+        end
+
+        context "when the election's parameters are not blocked" do
+          let(:model) { create(:election) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_election_type_spec.rb:93]:
```suggestion
          let(:model) { create(:vocdoni_election) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+module Decidim
+  module Vocdoni
+    describe VocdoniElectionsType, type: :graphql do
+      include_context "with a graphql class type"
+      let(:model) { create(:vocdoni_component) }
+
+      it_behaves_like "a component query type"
+
+      describe "elections" do
+        let!(:component_vocdoni) { create_list(:election, 2, :published, component: model) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:16]:
```suggestion
        let!(:component_vocdoni) { create_list(:vocdoni_election, 2, :published, component: model) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        let!(:component_vocdoni_hidden) { create_list(:election, 2, component: model) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:17]:
```suggestion
        let!(:component_vocdoni_hidden) { create_list(:vocdoni_election, 2, component: model) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        let!(:other_elections) { create_list(:election, 2) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:18]:
```suggestion
        let!(:other_elections) { create_list(:vocdoni_election, 2) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+          expect(ids).not_to include(*other_elections.map(&:id).map(&:to_s))
+        end
+      end
+
+      describe "election" do
+        let(:query) { "query VocdoniElection($id: ID!){ election(id: $id) { id } }" }
+        let(:variables) { { id: election.id.to_s } }
+
+        context "when the election belongs to the component" do
+          let!(:election) { create(:election, :published, component: model) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:35]:
```suggestion
          let!(:election) { create(:vocdoni_election, :published, component: model) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+          it "finds the election" do
+            expect(response["election"]["id"]).to eq(election.id.to_s)
+          end
+        end
+
+        context "when the election doesn't belong to the component" do
+          let!(:election) { create(:election) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:43]:
```suggestion
          let!(:election) { create(:vocdoni_election) }
```

File: spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+          it "returns null" do
+            expect(response["election"]).to be_nil
+          end
+        end
+
+        context "when the election belongs to the component and its publication time is nil" do
+          let!(:election) { create(:election, component: model) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_elections_type_spec.rb:51]:
```suggestion
          let!(:election) { create(:vocdoni_election, component: model) }
```

File: spec/types/decidim/vocdoni/vocdoni_question_type_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      describe "weight" do
+        let(:query) { "{ weight }" }
+
+        it "returns the election weight" do
+          expect(response["weight"]).to eq(model.weight)
+        end
+      end
+
+      describe "answers" do
+        let!(:question2) { create(:question, :complete) }
```
Comment (andreslucena) [inline spec/types/decidim/vocdoni/vocdoni_question_type_spec.rb:43]:
```suggestion
        let!(:question2) { create(:vocdoni_question, :complete) }
```

**Comment by andreslucena** [comment]
@microstudi this is ready to reviewed again.

**Review by microstudi — APPROVED** [review]
super!

**Pull request merged**